source/tutorials/integration-testing-using-virtual-machines.rst

```diff
@@ -0,0 +1,228 @@
+Integration testing using Virtual Machines
+==========================================
+
+One of the most powerful features in the Nix ecosystem is **the ability
+to provide a set of declarative NixOS configurations and use a simple
+Python interface** to interact with them using `QEMU <https://www.qemu.org/>`_
+as the backend.
+
+They are also known as **NixOS tests**.
+
+Integration tests are reproducible due to the design properties of Nix,
+making them a valuable part of a Continuous Integration pipeline.
+
+.. note:: 
+
+  There is work in progress to `run integration tests on 
+  macOS host <https://github.com/NixOS/nixpkgs/issues/108984>`_.
+
+
+Testing a typical web application backed by PostgreSQL
+------------------------------------------------------
+
+This tutorial follows `PostgREST tutorial <https://postgrest.org/en/stable/tutorials/tut0.html>`_,
+a generic `RESTful API <https://restfulapi.net/>`_ for PostgreSQL.
+
+If you skim over the official tutorial, you'll notice there's quite a bit of setup
+in order to test if all the steps work.
+
+We are going to set up:
+
+- VM named ``server`` running postgreSQL and postgREST.
+
+- VM named ``client`` running HTTP client queries using ``curl``.
+  
+- A ``testScript`` orchestrating testing logic between ``client`` and ``server``.
+
+
+Writing the test
+----------------
+
+Create ``postgrest.nix``:
+
+.. TODO: highlight nix https://github.com/pygments/pygments/issues/1793
+
+
+.. code-block::
+  :linenos:
+
+  let
+    # Pin nixpkgs, see pinning tutorial for more details
+    nixpkgs = fetchTarball "https://github.com/NixOS/nixpkgs/archive/0f8f64b54ed07966b83db2f20c888d5e035012ef.tar.gz";
+
+    # Single source of truth for all tutorial constants
+    database      = "postgres";
+    schema        = "api";
+    table         = "todos";
+    username      = "authenticator";
+    password      = "mysecretpassword";
+    webRole       = "web_anon";
+    postgrestPort = 3000;
+
+    # NixOS module shared between server and client
+    sharedModule = {
+      # don't assume graphical interface is possible
+      virtualisation.graphics = false;
```

Comment:
Why is this necessary? Does it speed up the test? If so, I should add that to my own tests.

Reply:
It makes sure the tests works on CI where there's no $DISPLAY.

Reply:
This one caught me off-guard, too - how about phrasing such as this?

> since it's common for CI not to have $DISPLAY available, we have to explicitly tell the tests "please don't expect any screen available"

(this concrete wording is kinda flake-y, but hopefully will work as a sales pitch here 🙂)